pkg/diagnostics/utils/trace_utils.go

```diff
@@ -54,29 +69,74 @@ func GetTraceSamplingRate(rate string) float64 {
 }
 
 // TraceSampler returns Probability Sampler option.
-func TraceSampler(samplingRate string) trace.StartOption {
-	return trace.WithSampler(trace.ProbabilitySampler(GetTraceSamplingRate(samplingRate)))
+func TraceSampler(samplingRate string) sdktrace.Sampler {
+	return sdktrace.ParentBased(sdktrace.TraceIDRatioBased(GetTraceSamplingRate(samplingRate)))
 }
 
 // IsTracingEnabled parses the given rate and returns false if sampling rate is explicitly set 0.
 func IsTracingEnabled(rate string) bool {
 	return GetTraceSamplingRate(rate) != 0
 }
 
-// SpanFromContext returns the SpanContext stored in a context, or nil if there isn't one.
+// SpanFromContext returns the SpanContext stored in a context, or nil or trace.nooSpan{} if there isn't one. - TODO
 func SpanFromContext(ctx context.Context) *trace.Span {
 	if reqCtx, ok := ctx.(*fasthttp.RequestCtx); ok {
 		val := reqCtx.UserValue(daprFastHTTPContextKey)
-		if val == nil {
-			return nil
+		if val != nil {
+			return val.(*trace.Span)
 		}
-		return val.(*trace.Span)
 	}
 
-	return trace.FromContext(ctx)
+	span := trace.SpanFromContext(ctx)
+	return &span
 }
 
 // SpanToFastHTTPContext sets span into fasthttp.RequestCtx.
 func SpanToFastHTTPContext(ctx *fasthttp.RequestCtx, span *trace.Span) {
 	ctx.SetUserValue(daprFastHTTPContextKey, span)
 }
+
+// BinaryFromSpanContext returns the binary format representation of a SpanContext.
+//
+// If sc is the zero value, Binary returns nil.
+func BinaryFromSpanContext(sc trace.SpanContext) []byte {
+	traceID := sc.TraceID()
+	spanID := sc.SpanID()
+	traceFlags := sc.TraceFlags()
+	if sc.Equal(trace.SpanContext{}) {
+		return nil
+	}
+	var b [29]byte
+	copy(b[2:18], traceID[:])
+	b[18] = 1
```

Comment:
Where are these magic numbers coming from?

Reply:
That comes from here: https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/trace/propagation/propagation.go#L57

Previous tracing implementation used that function from opencensus. I had to create my own function to be able to remove all references to opencensus.

Reply:
It would be great if you could add a comment saying where that code comes from, and report its copyright. Something like:

```go
// Adapted from https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/trace/propagation/propagation.go
// Copyright 2017 OpenCensus Authors
// License: https://github.com/census-instrumentation/opencensus-go/blob/49838f207d61097fc0ebb8aeef306913388376ca/LICENSE (Apache-2.0)
```